### PR TITLE
Deletion

### PR DESCRIPTION
**Pull request by thehabes**

### CODE REVIEW

File: src/java/edu/slu/action/ObjectAction.java [hunk 1]
```diff
@@ -811,28 +843,222 @@ public void updateObject() throws IOException, ServletException, Exception{
...
+         if(detectedPrevious){ 
+             //The object being deleted had a previous.  That previous object next[] must be updated with the deleted object's next[].
+             BasicDBObject query2 = new BasicDBObject();
+             BasicDBObject objToUpdate2;
+             BasicDBObject objWithUpdate2;
+             query2.append("@id", previous_id);
+             objToUpdate2 = (BasicDBObject) mongoDBService.findOneByExample(Constant.COLLECTION_ANNOTATION, query2); 
+             if(null != objToUpdate2){
+                JSONObject fixHistory2 = JSONObject.fromObject(objToUpdate2);  
+                fixHistory2.getJSONObject("__rerum").getJSONObject("history").element("next", next_ids);
```
Comment (cubap) [inline src/java/edu/slu/action/ObjectAction.java]: What needs to happen is:

fixHistory2.__rerum.history.next needs to have the DO.@id sliced out and the DO.__rerum.history.next array concatenated.

File: src/java/edu/slu/action/ObjectAction.java [hunk 2]
```diff
@@ -811,28 +843,222 @@ public void updateObject() throws IOException, ServletException, Exception{
...
+         for(int n=0; n<next_ids.size(); n++){
+             BasicDBObject query = new BasicDBObject();
+             BasicDBObject objToUpdate;
+             BasicDBObject objWithUpdate;
+             String nextID = next_ids.getString(n);
+             query.append("@id", nextID);
+             objToUpdate = (BasicDBObject) mongoDBService.findOneByExample(Constant.COLLECTION_ANNOTATION, query); 
+             if(null != objToUpdate){
+                JSONObject fixHistory = JSONObject.fromObject(objToUpdate);
+                if(isRoot){ //The object being deleted was root.  That means these next objects must become root.  Strictly, all history trees must have num(root) > 0.
```
[the remark this answers is not in the document]
Comment (thehabes) [reply src/java/edu/slu/action/ObjectAction.java]: When you delete a prime, there's potential to make more than one prime.  getAllVersions can return more than one prime for a history tree (AKA there can be more than one node on level 0.  From level 0 down, each branch will have one prime=root, but at level 0 where the branches start the potential is infinite).

File: src/java/edu/slu/action/ObjectAction.java [hunk 3]
```diff
@@ -811,28 +843,222 @@ public void updateObject() throws IOException, ServletException, Exception{
...
+     * 
+     * @param deletedObj A JSONObject of the object being deleted.
+     * @return A boolean representing whether or not this function succeeded. 
+     */
+     public boolean greenThumb(JSONObject obj){
+         boolean success = true;
+         String previous_id = obj.getJSONObject("__rerum").getJSONObject("history").getString("previous");
+         String prime_id = obj.getJSONObject("__rerum").getJSONObject("history").getString("prime");
+         JSONArray next_ids = obj.getJSONObject("__rerum").getJSONObject("history").getJSONArray("next");
+         //@cubap We are operating under the assumption the object MUST HAVE these values.  This will throw an error if not.  Should I handle with a try-catch?
```
[the remark this answers is not in the document]
Comment (thehabes) [reply src/java/edu/slu/action/ObjectAction.java]: Maybe the response to this issue should be a checkVersion() helper function we can react to in these spots in actions.

File: src/java/edu/slu/action/ObjectAction.java [hunk 4]
```diff
@@ -811,28 +843,222 @@ public void updateObject() throws IOException, ServletException, Exception{
...
+                            //@cubap I don't know if this is what we want to say if this happens...
+                            writeErrorResponse("We could not update the history tree correctly.  The delete failed.", HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+                        }
+                    }
+                    else{
+                        writeErrorResponse("The id string provided for DELETE could not be found in RERUM: "+received.getString("@id")+". \n DELETE failed.", HttpServletResponse.SC_NOT_FOUND);
+                    }
+                }
+                else{
+                    writeErrorResponse("Object for delete did not contain an id.  Could not delete.", HttpServletResponse.SC_BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (thehabes) [reply src/java/edu/slu/action/ObjectAction.java]: yea

File: src/java/edu/slu/action/ObjectAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cubap) [reply src/java/edu/slu/action/ObjectAction.java]: getAllVersions() will only return one prime node. Inside the \[next] loop, each getDescendants() only has one prime at a time.

Ultimately, I think this conversation doesn't effect anything though.

File: src/java/edu/slu/action/ObjectAction.java [hunk 5]
```diff
@@ -811,28 +843,222 @@ public void updateObject() throws IOException, ServletException, Exception{
...
+                    newTreePrime(fixHistory);
+                    //Do not overwrite previous of the new prime objects so they reference the deleted prime object for the future.
+                    //fixHistory.getJSONObject("__rerum").getJSONObject("history").element("previous", "");
+                }
+                else if(detectedPrevious){ //The object being deleted had a previous.  That is now absorbed by this next object to mend the gap.  
+                    fixHistory.getJSONObject("__rerum").getJSONObject("history").element("previous", previous_id);
+                }
+                else{
+                    //Yikes this is some kind of error...it is either root or has a previous, this case means neither are true.
+                    success = false;
```
[the remark this answers is not in the document]
Comment (thehabes) [reply src/java/edu/slu/action/ObjectAction.java]: I wonder about the implications on the objects that reference it but ultimately think we should allow it to delete still.

**Pull request merged**